File: src/main/java/org/stt/reporting/WorkingtimeItemProvider.java
package org.stt.reporting;

import org.stt.config.WorktimeConfig;
import org.stt.time.DateTimes;

import javax.inject.Inject;
import javax.inject.Named;
import java.io.*;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.time.LocalDate;
import java.util.HashMap;
import java.util.Map;
import java.util.TreeMap;
import java.util.logging.Logger;

import static java.util.Objects.requireNonNull;

/**
 * Reads information about working times from the configured workingTimes file
 * and aggregates them into {@link WorkingtimeItem}s
 */
public class WorkingtimeItemProvider {
    private static final Logger LOG = Logger.getLogger(WorkingtimeItemProvider.class.getSimpleName());
    private final WorktimeConfig config;
    private Map<LocalDate, WorkingtimeItem> workingHoursPerDay = new HashMap<>();

	@Inject
    public WorkingtimeItemProvider(WorktimeConfig config,
                                   @Named("homePath") String homePath) {
        this.config = requireNonNull(config);

        File workingTimesFile = config.getWorkingTimesFile().file(homePath);
        if (workingTimesFile.exists()) {
            populateHoursMapsFromFile(workingTimesFile);
        }
    }

	/**
	 * 
	 * @return the dates and corresponding absence times
	 */
    public Map<LocalDate, WorkingtimeItem> getOvertimeAbsences() {
        Map<LocalDate, WorkingtimeItem> overtimeAbsenceDuration = new TreeMap<>();
        for (Map.Entry<LocalDate, WorkingtimeItem> e : workingHoursPerDay
                .entrySet()) {
			// if time is negative...
            if (e.getValue().getMin().isNegative()) {
                overtimeAbsenceDuration.put(e.getKey(), e.getValue());
			}
		}

		return overtimeAbsenceDuration;
	}

	/**
	 * 
	 * @param date
	 * @return the configured duration to be worked without producing positive
	 *         or negative overtime
	 */
	public WorkingtimeItem getWorkingTimeFor(LocalDate date) {
        WorkingtimeItem workingHours = workingHoursPerDay.get(date);
        if (workingHours == null) {
            return fromDuration(config.getWorkingHours().getOrDefault(date.getDayOfWeek(), Duration.ZERO));
        }
        return workingHours;
    }

	private void populateHoursMapsFromFile(File workingTimesFile) {

		try (BufferedReader wtReader = new BufferedReader(
				new InputStreamReader(constructReaderFrom(workingTimesFile), StandardCharsets.UTF_8))) {

			String currentLine;
			while ((currentLine = wtReader.readLine()) != null) {
				if (currentLine.matches("^\\d+.*")) {
					// it is a date
					String[] split = currentLine.split("\\s+");
                    LocalDate parseDateTime = LocalDate.parse(split[0], DateTimes.DATE_TIME_FORMATTER_YYYY_MM_DD_DASHED);
                    String minHours = split[1];
                    String maxHours = minHours;

					if (split.length > 2) {
						maxHours = split[2];
					}
					workingHoursPerDay.put(parseDateTime,
							fromHours(minHours, maxHours));

				} else if (currentLine.startsWith("hours")) {
                    LOG.severe("'hours' is no longer supported in your worktimes file, please setup default working hours in your configuration.");
                }
            }
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
    }

	private InputStream constructReaderFrom(File workingTimesFile) throws FileNotFoundException {
		if(workingTimesFile.getName().equalsIgnoreCase("-")) {
			return System.in;
		}
		return new FileInputStream(workingTimesFile);
	}

	private WorkingtimeItem fromHours(String minHours, String maxHours) {
        Duration minDur = Duration.ofHours(Long.parseLong(minHours));
        Duration maxDur = Duration.ofHours(Long.parseLong(maxHours));
        return new WorkingtimeItem(minDur, maxDur);
	}

    private WorkingtimeItem fromDuration(Duration duration) {
        return new WorkingtimeItem(duration, duration);
    }

	/**
	 *  
	 */
	public static class WorkingtimeItem {

		private Duration min;
		private Duration max;

		public WorkingtimeItem(Duration min, Duration max) {
			this.min = min;
			this.max = max;
		}

		public Duration getMin() {
			return min;
		}

		public Duration getMax() {
			return max;
		}

		@Override
		public int hashCode() {
			final int prime = 31;
			int result = 1;
			result = prime * result + ((max == null) ? 0 : max.hashCode());
			result = prime * result + ((min == null) ? 0 : min.hashCode());
			return result;
		}

		@Override
		public boolean equals(Object obj) {
			if (this == obj) {
				return true;
			}
			if (obj == null) {
				return false;
			}
			if (getClass() != obj.getClass()) {
				return false;
			}
			WorkingtimeItem other = (WorkingtimeItem) obj;
			if (max == null) {
				if (other.max != null) {
					return false;
				}
			} else if (!max.equals(other.max)) {
				return false;
			}
			if (min == null) {
				if (other.min != null) {
					return false;
				}
			} else if (!min.equals(other.min)) {
				return false;
			}
			return true;
		}
	}
}

File: src/main/kotlin/org/stt/reporting/WorkingtimeItemProvider.kt
package org.stt.reporting

import org.stt.config.WorktimeConfig
import org.stt.reporting.WorkingtimeItemProvider.WorkingtimeItem
import org.stt.time.DateTimes
import java.io.*
import java.nio.charset.StandardCharsets
import java.time.Duration
import java.time.LocalDate
import java.util.*
import java.util.Objects.requireNonNull
import java.util.logging.Logger
import javax.inject.Inject
import javax.inject.Named

/**
 * Reads information about working times from the configured workingTimes file
 * and aggregates them into [WorkingtimeItem]s
 */
class WorkingtimeItemProvider @Inject
constructor(config: WorktimeConfig,
            @Named("homePath") homePath: String) {
    private val config: WorktimeConfig
    private val workingHoursPerDay = HashMap<LocalDate, WorkingtimeItem>()

    /**
     *
     * @return the dates and corresponding absence times
     */
    // if time is negative...
    val overtimeAbsences: Map<LocalDate, WorkingtimeItem>
        get() {
            val overtimeAbsenceDuration = TreeMap<LocalDate, WorkingtimeItem>()
            for ((key, value) in workingHoursPerDay) {
                if (value.min.isNegative) {
                    overtimeAbsenceDuration[key] = value
                }
            }

            return overtimeAbsenceDuration
        }

    init {
        this.config = requireNonNull(config)

        val workingTimesFile = config.workingTimesFile.file(homePath)
        if (workingTimesFile.exists()) {
            populateHoursMapsFromFile(workingTimesFile)
        }
    }

    /**
     *
     * @return the configured duration to be worked without producing positive
     * or negative overtime
     */
    fun getWorkingTimeFor(date: LocalDate): WorkingtimeItem {
        return workingHoursPerDay[date]
                ?: return fromDuration(config.workingHours.getOrDefault(date.dayOfWeek.name, Duration.ZERO))
    }

    private fun populateHoursMapsFromFile(workingTimesFile: File) {
        BufferedReader(InputStreamReader(constructReaderFrom(workingTimesFile), StandardCharsets.UTF_8))
                .forEachLine { currentLine ->
                    if (currentLine.matches("^\\d+.*".toRegex())) {
                        // it is a date
                        val split = currentLine.split("\\s+".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
                        val parseDateTime = LocalDate.parse(split[0], DateTimes.DATE_TIME_FORMATTER_YYYY_MM_DD_DASHED)
                        val minHours = split[1]
                        var maxHours = minHours

                        if (split.size > 2) {
                            maxHours = split[2]
                        }
                        workingHoursPerDay[parseDateTime] = fromHours(minHours, maxHours)

                    } else if (currentLine.startsWith("hours")) {
                        LOG.severe("'hours' is no longer supported in your worktimes file, please setup default working hours in your configuration.")
                    }
                }
    }

    @Throws(FileNotFoundException::class)
    private fun constructReaderFrom(workingTimesFile: File): InputStream {
        return if (workingTimesFile.name.equals("-", ignoreCase = true)) {
            System.`in`
        } else FileInputStream(workingTimesFile)
    }

    private fun fromHours(minHours: String, maxHours: String): WorkingtimeItem {
        val minDur = Duration.ofHours(java.lang.Long.parseLong(minHours))
        val maxDur = Duration.ofHours(java.lang.Long.parseLong(maxHours))
        return WorkingtimeItem(minDur, maxDur)
    }

    private fun fromDuration(duration: Duration): WorkingtimeItem {
        return WorkingtimeItem(duration, duration)
    }

    /**
     *
     */
    data class WorkingtimeItem(val min: Duration, val max: Duration)

    companion object {
        private val LOG = Logger.getLogger(WorkingtimeItemProvider::class.java.simpleName)
    }
}
